theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,686 @@
+
+    if (isinstance(node.op, AbstractConv2d) or
+        isinstance(node.op, AbstractConv2d_gradWeights) or
+        isinstance(node.op, AbstractConv2d_gradInputs)):
+        #conv(host_from_gpu) -> host_from_gpu(gpu_conv)
+        inp1 = node.inputs[0]
+        inp2 = node.inputs[1]
+        inp1_on_gpu = (inp1.owner and isinstance(inp1.owner.op, HostFromGpu))
+        inp2_on_gpu = (inp2.owner and isinstance(inp2.owner.op, HostFromGpu))
+
+        if inp1_on_gpu or inp2_on_gpu:
+            conv = node.op
+            inps = list(node.inputs)
+            inps[0] = gpu_from_host(inps[0])
+            inps[1] = gpu_from_host(inps[1])
+            out = conv(*inps)
+            out = theano.tensor.patternbroadcast(
+                out,
+                node.outputs[0].broadcastable)
+            out.values_eq_approx = values_eq_approx_high_tol
+            return [as_tensor_variable(out)]
+# We register the optimizer that moves convolutions to the GPU.
+#register_gpu()(local_conv2d_gpu_conv)
+
+
+
+### Call dnn conv class directly
+@local_optimizer([AbstractConv2d,
+                  AbstractConv2d_gradWeights,
+                  AbstractConv2d_gradInputs])
+def local_conv2d_cudnn(node):
+
+    inp1 = node.inputs[0]
+    inp2 = node.inputs[1]
+
+    if not isinstance(inp1.type, CudaNdarrayType) or \
+            not isinstance(inp2.type, CudaNdarrayType):
+        return None
+    if not dnn_available():
+        return None
+
+    if node.op.filter_flip:
+        conv_mode = 'conv'
+    else:
+        conv_mode = 'cross'
+    if (isinstance(node.op, AbstractConv2d)):
+        rval = dnn_conv(inp1, inp2,
+                        border_mode=node.op.border_mode,
+                        subsample=node.op.subsample,
+                        direction_hint='forward',
+                        conv_mode = conv_mode)
+        return [rval]
+    if (isinstance(node.op, AbstractConv2d_gradWeights)):
+        rval = dnn_conv(inp1.dimshuffle(1, 0, 2, 3), inp2,
+                        border_mode=node.op.border_mode,
+                        subsample=node.op.subsample,
+                        direction_hint='bprop weights',
+                        conv_mode = conv_mode)
+        return [rval]
+    if (isinstance(node.op, AbstractConv2d_gradInputs)):
+        rval = dnn_conv(inp1, inp2,
+                        border_mode=node.op.border_mode,
+                        subsample=node.op.subsample,
+                        direction_hint='bprop inputs',
+                        conv_mode = conv_mode)
+        return [rval]
+register_specialize_device(local_conv2d_cudnn)
+
+
+@local_optimizer([AbstractConv2d])
+def local_conv2d_corrmm(node):
+
+    img, kern = node.inputs
+    if (not isinstance(img.type, CudaNdarrayType) or
+            not isinstance(kern.type, CudaNdarrayType)):
+        return None
+
+    print "here"
+
+    if node.op.border_mode in ['full', 'valid']:
+        border_mode = node.op.border_mode
+        subsample = node.op.subsample
```

Comment:
I think we should flip the kernel (if necessary) once at the beginning, not in the branches.